File: tracking/src/kalman_filter.cpp
/*
kalman_filter.cpp

Constant-velocity Kalman filter for 2D tracking.
*/

#include "kalman_filter.hpp"

#include <iostream>


KalmanFilter2D::KalmanFilter2D(

    double dt,

    double process_noise,

    double measurement_noise
)
{
    this->dt = dt;

    this->process_noise =
        process_noise;

    this->measurement_noise =
        measurement_noise;

    this->x = {
        0.0,
        0.0,
        0.0,
        0.0
    };
}


void KalmanFilter2D::initialize(
    double x,
    double y
)
{
    this->x[0] = x;
    this->x[1] = y;

    this->x[2] = 0.0;
    this->x[3] = 0.0;
}


void KalmanFilter2D::predict()
{
    // ------------------------------------------------------------
    // Constant velocity prediction
    // ------------------------------------------------------------

    this->x[0] += this->x[2] * dt;

    this->x[1] += this->x[3] * dt;
}


void KalmanFilter2D::update(
    double measured_x,
    double measured_y
)
{
    // ------------------------------------------------------------
    // Residual
    // ------------------------------------------------------------

    double residual_x =
        measured_x - this->x[0];

    double residual_y =
        measured_y - this->x[1];

    // ------------------------------------------------------------
    // Simple smoothing gain
    // ------------------------------------------------------------

    double gain = 0.5;

    // ------------------------------------------------------------
    // Velocity update
    // ------------------------------------------------------------

    this->x[2] =
        residual_x / dt;

    this->x[3] =
        residual_y / dt;

    // ------------------------------------------------------------
    // Position update
    // ------------------------------------------------------------

    this->x[0] +=
        gain * residual_x;

    this->x[1] +=
        gain * residual_y;
}


double KalmanFilter2D::getX() const
{
    return this->x[0];
}


double KalmanFilter2D::getY() const
{
    return this->x[1];
}


double KalmanFilter2D::getVX() const
{
    return this->x[2];
}


double KalmanFilter2D::getVY() const
{
    return this->x[3];
}

File: tracking/state_estimation/kalman_filter.py
"""
kalman_filter.py

Constant-velocity Kalman filter for 2D object tracking.

Purpose:
- Predict object motion
- Smooth noisy detections
- Stabilize temporal tracking
- Provide motion-aware state estimation

State Vector
------------
[x, y, vx, vy]

where:
x  -> center x
y  -> center y
vx -> velocity x
vy -> velocity y
"""

import numpy as np


class KalmanFilter2D:

    def __init__(
        self,
        dt=1.0,
        process_noise=1e-2,
        measurement_noise=1e-1
    ):

        self.dt = dt

        # --------------------------------------------------------
        # State vector
        # --------------------------------------------------------
        # [x, y, vx, vy]
        # --------------------------------------------------------

        self.x = np.zeros(
            (4, 1),
            dtype=np.float64
        )

        # --------------------------------------------------------
        # State transition matrix
        # --------------------------------------------------------

        self.F = np.array([

            [1, 0, dt, 0],

            [0, 1, 0, dt],

            [0, 0, 1, 0],

            [0, 0, 0, 1]

        ], dtype=np.float64)

        # --------------------------------------------------------
        # Measurement matrix
        # --------------------------------------------------------
        # We only observe:
        # [x, y]
        # --------------------------------------------------------

        self.H = np.array([

            [1, 0, 0, 0],

            [0, 1, 0, 0]

        ], dtype=np.float64)

        # --------------------------------------------------------
        # State covariance
        # --------------------------------------------------------

        self.P = np.eye(
            4,
            dtype=np.float64
        ) * 100.0

        # --------------------------------------------------------
        # Process noise
        # --------------------------------------------------------

        self.Q = np.eye(
            4,
            dtype=np.float64
        ) * process_noise

        # --------------------------------------------------------
        # Measurement noise
        # --------------------------------------------------------

        self.R = np.eye(
            2,
            dtype=np.float64
        ) * measurement_noise

        # --------------------------------------------------------
        # Identity matrix
        # --------------------------------------------------------

        self.I = np.eye(
            4,
            dtype=np.float64
        )

    def initialize(
        self,
        x,
        y
    ):
        """
        Initialize filter state.
        """

        self.x = np.array([

            [x],
            [y],
            [0.0],
            [0.0]

        ], dtype=np.float64)

    def predict(self):
        """
        Predict next state.
        """

        # --------------------------------------------------------
        # State prediction
        # --------------------------------------------------------

        self.x = self.F @ self.x

        # --------------------------------------------------------
        # Covariance prediction
        # --------------------------------------------------------

        self.P = (

            self.F @ self.P @ self.F.T
            + self.Q
        )

        return self.get_state()

    def update(
        self,
        measured_x,
        measured_y
    ):
        """
        Update filter using measurement.
        """

        z = np.array([

            [measured_x],
            [measured_y]

        ], dtype=np.float64)

        # --------------------------------------------------------
        # Innovation
        # --------------------------------------------------------

        y = z - (self.H @ self.x)

        # --------------------------------------------------------
        # Innovation covariance
        # --------------------------------------------------------

        S = (

            self.H @ self.P @ self.H.T
            + self.R
        )

        # --------------------------------------------------------
        # Kalman gain
        # --------------------------------------------------------

        K = (

            self.P @ self.H.T @
            np.linalg.inv(S)
        )

        # --------------------------------------------------------
        # State update
        # --------------------------------------------------------

        self.x = self.x + (K @ y)

        # --------------------------------------------------------
        # Covariance update
        # --------------------------------------------------------

        self.P = (

            self.I - (K @ self.H)
        ) @ self.P

        return self.get_state()

    def get_state(self):
        """
        Return current estimated state.
        """

        return {

            "x": float(self.x[0, 0]),

            "y": float(self.x[1, 0]),

            "vx": float(self.x[2, 0]),

            "vy": float(self.x[3, 0])
        }


def main():

    print("\n=== Kalman Filter Validation ===\n")

    kf = KalmanFilter2D()

    kf.initialize(
        x=100,
        y=100
    )

    measurements = [

        (105, 102),
        (111, 105),
        (118, 109),
        (126, 114),
        (135, 120)
    ]

    for i, (mx, my) in enumerate(measurements):

        prediction = kf.predict()

        updated = kf.update(
            mx,
            my
        )

        print(
            f"\nStep {i}"
        )

        print(
            f"Measurement: ({mx}, {my})"
        )

        print(
            f"Prediction: "
            f"({prediction['x']:.2f}, "
            f"{prediction['y']:.2f})"
        )

        print(
            f"Updated: "
            f"({updated['x']:.2f}, "
            f"{updated['y']:.2f})"
        )

        print(
            f"Velocity: "
            f"({updated['vx']:.2f}, "
            f"{updated['vy']:.2f})"
        )


if __name__ == "__main__":
    main()